4876f3fcfcda2720d2ff0cb99cf7c27d490e27c9 osdep/linux/hostdisk: Use stat() instead of udevadm for partition lookup
 grub-core/osdep/linux/hostdisk.c | 52 ++++----------------------------
 1 file changed, 6 insertions(+), 46 deletions(-)

diff --git a/grub-core/osdep/linux/hostdisk.c b/grub-core/osdep/linux/hostdisk.c
index da62f924e..d3326d095 100644
--- a/grub-core/osdep/linux/hostdisk.c
+++ b/grub-core/osdep/linux/hostdisk.c
@@ -31,6 +31,7 @@
 #include <grub/misc.h>
 #include <grub/i18n.h>
 #include <grub/list.h>
+#include <grub/osdep/major.h>
 
 #include <stdio.h>
 #include <stdlib.h>
@@ -98,54 +99,13 @@ grub_util_get_fd_size_os (grub_util_fd_t fd, const char *name, unsigned *log_sec
 static char *
 sysfs_partition_path (const char *dev, const char *entry)
 {
-  const char *argv[7];
-  int fd;
-  pid_t pid;
-  FILE *udevadm;
-  char *buf = NULL;
-  size_t len = 0;
-  char *path = NULL;
-
-  argv[0] = "udevadm";
-  argv[1] = "info";
-  argv[2] = "--query";
-  argv[3] = "path";
-  argv[4] = "--name";
-  argv[5] = dev;
-  argv[6] = NULL;
-
-  pid = grub_util_exec_pipe (argv, &fd);
-
-  if (!pid)
-    return NULL;
-
-  /* Parent.  Read udevadm's output.  */
-  udevadm = fdopen (fd, "r");
-  if (!udevadm)
-    {
-      grub_util_warn (_("Unable to open stream from %s: %s"),
-		      "udevadm", strerror (errno));
-      close (fd);
-      goto out;
-    }
-
-  if (getline (&buf, &len, udevadm) > 0)
-    {
-      char *newline;
-
-      newline = strchr (buf, '\n');
-      if (newline)
-	*newline = '\0';
-      path = xasprintf ("/sys%s/%s", buf, entry);
-    }
+  struct stat st;
 
-out:
-  if (udevadm)
-    fclose (udevadm);
-  waitpid (pid, NULL, 0);
-  free (buf);
+  if (stat (dev, &st) == 0 && S_ISBLK (st.st_mode))
+    return xasprintf ("/sys/dev/block/%u:%u/%s",
+		      major (st.st_rdev), minor (st.st_rdev), entry);
 
-  return path;
+  return NULL;
 }
 
 static int
-- 
2.39.2

